FLINK-3930 Added shared secret based authorization for Flink service …

vijikarthi opened the pull request:
This PR addresses FLINK-3930 requirements. It enables shared secret based secure cookie authorization for the following components
- Akka layer
- Blob Service
- Web UI

Secure cookie authentication can be enabled by providing below configurations to Flink configuration file.
- `security.enabled`: A boolean value (true|false) indicating security is enabled or not.
- `security.cookie` : Secure cookie value to be used for authentication. For standalone deployment mode, the secure cookie value is mandatory when security is enabled but for the Yarn mode it is optional (auto-generated if not provided).

Alternatively, secure cookie value can be provided through Flink/Yarn CLI using "-k" or "--cookie" parameter option.

The web runtime module prompts for secure cookie using standard basic HTTP authentication mechanism, where the user id field is a noop and the password field will be used to capture the secure cookie.

vijikarthi commented:
@mxm - The patch is available for your review. Please take a look.

rmetzger commented:
Thank you for opening a pull request for this. I'll check it out now.

rmetzger commented:
It seems that some of the YARN tests are failing because the jobmanager / application master is logging:

```
2016-08-27 08:48:54,858 INFO  org.apache.flink.yarn.YarnJobManager                          - ApplicatonMaster will shut down YARN session when job f56f54e7288172fdc16966c04e343d91 has finished.
2016-08-27 08:48:55,370 ERROR org.apache.flink.runtime.blob.BlobServerConnection            - Error while executing BLOB connection.
java.io.EOFException: Read an incomplete length
	at org.apache.flink.runtime.blob.BlobUtils.readLength(BlobUtils.java:282)
	at org.apache.flink.runtime.blob.BlobServerConnection.readCookie(BlobServerConnection.java:487)
	at org.apache.flink.runtime.blob.BlobServerConnection.validateSecureCookie(BlobServerConnection.java:499)
	at org.apache.flink.runtime.blob.BlobServerConnection.run(BlobServerConnection.java:107)
2016-08-27 08:48:55,453 INFO  org.apache.flink.yarn.YarnJobManager                          - Submitting job f56f54e7288172fdc16966c04e343d91 (Flink Java Job at Sat Aug 27 08:48:46 UTC 2016).
2016-08-27 08:48:55,465 INFO  org.apache.flink.yarn.YarnJobManager                          - Using restart strategy NoRestartStrategy for f56f54e7288172fdc16966c04e343d91.
2016-08-27 08:48:55,524 INFO  org.apache.flink.yarn.YarnJobManager                          - Scheduling job f56f54e7288172fdc16966c04e343d91 (Flink Java Job at Sat Aug 27 08:48:46 UTC 2016).
2016-08-27 08:48:55,525 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph        - CHAIN DataSource (at getDefaultTextLineDataSet(WordCountData.java:70) (org.apache.flink.api.java.io.CollectionInputFormat)) -> FlatMap (FlatMap at main(WordCount.java:80)) -> Combine(SUM(1), at main(WordCount.java:83) (1/1) (f5b14d29193f2a64970a98423aa00be2) switched from CREATED to SCHEDULED
2016-08-27 08:48:55,535 INFO  org.apache.flink.yarn.YarnJobManager                          - Status of job f56f54e7288172fdc16966c04e343d91 (Flink Java Job at Sat Aug 27 08:48:46 UTC 2016) changed to RUNNING.
2016-08-27 08:48:55,547 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph        - CHAIN DataSource (at getDefaultTextLineDataSet(WordCountData.java:70) (org.apache.flink.api.java.io.CollectionInputFormat)) -> FlatMap (FlatMap at main(WordCount.java:80)) -> Combine(SUM(1), at main(WordCount.java:83) (1/1) (f5b14d29193f2a64970a98423aa00be2) switched from SCHEDULED to DEPLOYING
2016-08-27 08:48:55,547 INFO  org.apache.flink.runtime.executiongraph.ExecutionGraph        - Deploying CHAIN DataSource (at getDefaultTextLineDataSet(WordCountData.java:70) (org.apache.flink.api.java.io.CollectionInputFormat)) -> FlatMap (FlatMap at main(WordCount.java:80)) -> Combine(SUM(1), at main(WordCount.java:83) (1/1) (attempt #0) to testing-worker-linux-docker-8659bf87-3480-linux-9
2016-08-27 08:48:55,673 ERROR org.apache.flink.runtime.blob.BlobServerConnection            - Error while executing BLOB connection.
java.io.EOFException: Read an incomplete length
	at org.apache.flink.runtime.blob.BlobUtils.readLength(BlobUtils.java:282)
	at org.apache.flink.runtime.blob.BlobServerConnection.readCookie(BlobServerConnection.java:487)
	at org.apache.flink.runtime.blob.BlobServerConnection.validateSecureCookie(BlobServerConnection.java:499)
	at org.apache.flink.runtime.blob.BlobServerConnection.run(BlobServerConnection.java:107)
```
(see: https://s3.amazonaws.com/archive.travis-ci.org/jobs/155437254/log.txt)

rmetzger commented on docs/setup/config.md:
```diff
@@ -107,6 +107,21 @@ Please make sure to set the maximum ticket life span high long running jobs. The
+
+Alternatively, secure cookie value can be provided through Flink/Yarn CLI using "-k" or "--cookie" parameter option.
+
+The web runtime module prompts for secure cookie using standard basic HTTP authentication mechanism, where the user id field is noop and the password field will be used to capture the secure cookie.
```
Much appreciated that your pull request also contains documentation updates! Thank you.

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java:
```diff
@@ -77,7 +88,26 @@ public StandaloneClusterClient createCluster(
+
+		if(securityEnabled && secureCookieArg != null) {
+			LOG.debug("Secure cookie is provided as CLI argument and will be used");
+			config.setBoolean(ConfigConstants.SECURITY_ENABLED, true);
```
I think this line is not needed. if we enter the branch, `SECURITY_ENABLED` or `DEFAULT_SECURITY_ENABLED` is `true`.

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/cli/DefaultCLI.java:
```diff
@@ -77,7 +88,26 @@ public StandaloneClusterClient createCluster(
+	private void populateSecureCookieConfigurations(Configuration config, String secureCookieArg) {
+
+		boolean securityEnabled = config.getBoolean(ConfigConstants.SECURITY_ENABLED, DEFAULT_SECURITY_ENABLED);
+		LOG.debug("Security Enabled ? {}", securityEnabled);
```
(optional comment)
maybe its a bit nicer to call
```
if(securityEnabled) {
  LOG.debug("Security enabled");
} else {
  LOG.debug("Security disabled");
}
```

rmetzger commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/HttpRequestHandler.java:
```diff
@@ -99,7 +110,43 @@ public void channelRead0(ChannelHandlerContext ctx, HttpObject msg) {
 				}
-				
+
+				if(secureCookie != null) {
```
Why do you decide here if security is enabled based on the COOKIE's existence. I think it could happen that users torn security off, but still have some cookie set.

vijikarthi replied:
The secure cookie value could be auto-populated (Yarn) or user-provided but will be persisted in the in-memory Flink configuration instance which is passed to the web layer during bootstrap. Should the user decide to torn security off, then we expect the services to be restarted to reflect the change?

rmetzger replied:
The YARN code does the auto-populaton only if security is enabled, so it should not be an issue.

> user-provided but will be persisted in the in-memory Flink configuration instance which is passed to the web layer during bootstrap.

The Configuration object in the CliFrontend is not transferred to the ApplicationMaster / JobMaster.
The flink-config.yaml (and on YARN the "dynamic properties") are the only sources of the configuration for the AM/JM.

rmetzger commented on flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JobManagerConfigHandler.java:
```diff
@@ -40,12 +41,22 @@ public JobManagerConfigHandler(Configuration config) {
+			}
+			else {
+				gen.writeStringField("value", config.getString(key, null));
+			}
```
Good idea not to expose the setting here!
The `GlobalConfiguration.loadYAMLResource()` method is logging all the key / value pairs. Maybe it makes sense to hide the value there as well.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java:
```diff
@@ -687,7 +709,17 @@ else if (response != RETURN_OKAY) {
+				if(cookie instanceof String) {
+					secureCookie = (String) cookie;
+				}
+				LOG.info("Secure Cookie from JM: {}", secureCookie);
```
Why are you writing the cookie to the log file here?

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -97,6 +101,16 @@ public BlobServer(Configuration config) throws IOException {
+				throw new IllegalConfigurationException(errorMessage);
+			}
+		}
+
```
This looks a lot like duplicate code. Maybe it makes sense to move it to the `BlobUtils`, or even to the `SecurityUtils` as a static method, ensuring that security is enabled.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -426,4 +440,11 @@ void unregisterConnection(BlobServerConnection conn) {
+	public String getSecureCookie() { return secureCookie; }
+
+	/* Flag to indicate if service level authentication is enabled or not */
+	public boolean isSecurityEnabled() { return secureCookie != null; }
```
I think we can not decide whether security is enabled only based on the secure cookie's value.

vijikarthi replied:
We expect the secure cookie configuration to be available if security is enabled. A missing value will be reported as an error ahead of time. Are you expecting any other conditions to be met? Could you please elaborate?

rmetzger replied:
Okay, I see. In the constructor you are checking that the secureCookie is set if security is enabled.

Then its okay to leave it as is.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/JobGraph.java:
```diff
@@ -530,6 +529,7 @@ public void uploadRequiredJarFiles(InetSocketAddress serverAddress) throws IOExc
 			}
 		}
 	}
+	*/
```
I think its better to completely delete the method instead of commenting it out.

rmetzger commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -2041,6 +2044,18 @@ object JobManager {
+        LOG.error(message)
+        throw new RuntimeException(message)
+      }
+    }
```
As I said before, this is duplicate code that can be encapsulated into a `SecurityUtils` static method.

rmetzger commented on flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala:
```diff
@@ -1470,6 +1470,18 @@ object TaskManager {
+        throw new RuntimeException(message)
+      }
+    }
+
```
Duplicate code

rmetzger commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSecureTest.java:
```diff
@@ -0,0 +1,46 @@
+
+import static org.junit.Assert.fail;
+
+public class BlobClientSecureTest extends BlobClientTest {
```
Why are tests like the `testContentAddressableBuffer()` not failing, even if they pass a `null` security cookie?

rmetzger commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSecureTest.java:
```diff
@@ -0,0 +1,46 @@
+		catch (IOException e) {
+			e.printStackTrace();
+			fail(e.getMessage());
+		}
```
Maybe it makes sense to add a test ensuring that clients with a wrong cookie are rejected:

```
	@Test(expected = IOException.class)
	public void testInvalidSecurityCookie() throws IOException {

		BlobClient client = null;

		try {
			byte[] testBuffer = createTestBuffer();
			MessageDigest md = BlobUtils.createMessageDigest();
			md.update(testBuffer);

			InetSocketAddress serverAddress = new InetSocketAddress("localhost", BLOB_SERVER.getPort());
			client = new BlobClient(serverAddress, "different");

			// Store some data
			client.put(testBuffer);
		}
		finally {
			if (client != null) {
				try {
					client.close();
				} catch (Throwable t) {}
			}
		}
	}
```

vijikarthi replied:
Thanks. Will add the test case.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnApplicationMasterRunner.java:
```diff
@@ -597,6 +610,11 @@ public static ContainerLaunchContext createTaskManagerContext(
 
+		final String secureCookie = ENV.get(YarnConfigKeys.ENV_SECURE_AUTH_COOKIE);
+		if(secureCookie != null) {
+			containerEnv.put(YarnConfigKeys.ENV_SECURE_AUTH_COOKIE, secureCookie);
```
The problem here is that the secure cookie will be put into the environment of the TaskManager JVM, so it'll be quite easy to just read the environment variables (not sure if that is an issue).
Another issue is that YARN is by default launching processes by creating a temporary bash file, with all the environment variables and the JVM invocation. So the secure cookie will be written into some tmp directory on YARN.
I wonder if there's some infrastructure in YARN to transfer the tokens in a secure way.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -439,8 +450,8 @@ public static void runInteractiveCli(YarnClusterClient yarnCluster, boolean read
 						case "quit":
 						case "stop":
 							yarnCluster.shutdownCluster();
+							if (yarnCluster.hasBeenShutdown()) { removeAppState(applicationId); }
```
This is not consistent with the style in the rest of the Flink codebase. Can you put the "remoteAppState()" into a separate line?

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -682,6 +774,91 @@ public static File getYarnPropertiesLocation(Configuration conf) {
 	}
 
+	public static void persistAppState(String appId, String cookie) {
+		if(appId == null || cookie == null) { return; }
```
please put `return` into a new line.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -682,6 +774,91 @@ public static File getYarnPropertiesLocation(Configuration conf) {
+			LOG.warn("Exception occurred while fetching cookie for app id: {} Exception: {}", appId, e);
+		}
+	}
+
```
According to the design doc, the file is expected to be stored with special permissions. Where are you setting those?

vijikarthi replied:
We don't set any additional permissions since the file is managed in the user home directory.

rmetzger replied:
Okay, it'll rely on the default `umask` of the home directory.

I think it would make sense to change the permissions so that only the user can read/write the file, using this method: http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#setPosixFilePermissions%28java.nio.file.Path,%20java.util.Set%29

rmetzger commented:
I manually tested the code. Taskmanagers are properly rejected on missmatching cookies, it works when they match.

One thing I found was that the error reporting is not very good:

```
robert@robert-da ...k-1.2-SNAPSHOT-bin/flink-1.2-SNAPSHOT (git)-[FLINK-3930] % ./bin/flink run -k asdf ./examples/batch/WordCount.jar
Cluster configuration: Standalone cluster with JobManager at /127.0.0.1:6123
Using address 127.0.0.1:6123 to connect to JobManager.
JobManager web interface address http://127.0.0.1:8081
Starting execution of program
Executing WordCount example with default input data set.
Use --input to specify file input.
Printing result to stdout. Use --output to specify output path.
Submitting job with JobID: 0816b6a3427cdcfc1160655edc2cef09. Waiting for job completion.

------------------------------------------------------------
 The program finished with the following exception:

org.apache.flink.client.program.ProgramInvocationException: The program execution failed: Couldn't retrieve the JobExecutionResult from the JobManager.
	at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:404)
	at org.apache.flink.client.program.StandaloneClusterClient.submitJob(StandaloneClusterClient.java:92)
	at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:380)
	at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:367)
	at org.apache.flink.client.program.ContextEnvironment.execute(ContextEnvironment.java:61)
	at org.apache.flink.api.java.ExecutionEnvironment.execute(ExecutionEnvironment.java:896)
	at org.apache.flink.api.java.DataSet.collect(DataSet.java:410)
	at org.apache.flink.api.java.DataSet.print(DataSet.java:1605)
	at org.apache.flink.examples.java.wordcount.WordCount.main(WordCount.java:92)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:509)
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:403)
	at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:322)
	at org.apache.flink.client.CliFrontend.executeProgram(CliFrontend.java:775)
	at org.apache.flink.client.CliFrontend.run(CliFrontend.java:251)
	at org.apache.flink.client.CliFrontend.parseParameters(CliFrontend.java:1003)
	at org.apache.flink.client.CliFrontend.main(CliFrontend.java:1046)
Caused by: org.apache.flink.runtime.client.JobExecutionException: Couldn't retrieve the JobExecutionResult from the JobManager.
	at org.apache.flink.runtime.client.JobClient.awaitJobResult(JobClient.java:292)
	at org.apache.flink.runtime.client.JobClient.submitJobAndWait(JobClient.java:377)
	at org.apache.flink.client.program.ClusterClient.run(ClusterClient.java:400)
	... 19 more
Caused by: org.apache.flink.runtime.client.JobClientActorConnectionTimeoutException: Lost connection to the JobManager.
	at org.apache.flink.runtime.client.JobClientActor.handleMessage(JobClientActor.java:207)
	at org.apache.flink.runtime.akka.FlinkUntypedActor.handleLeaderSessionID(FlinkUntypedActor.java:90)
	at org.apache.flink.runtime.akka.FlinkUntypedActor.onReceive(FlinkUntypedActor.java:70)
	at akka.actor.UntypedActor$$anonfun$receive$1.applyOrElse(UntypedActor.scala:167)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:465)
	at akka.actor.UntypedActor.aroundReceive(UntypedActor.scala:97)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:516)
	at akka.actor.ActorCell.invoke(ActorCell.scala:487)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:254)
	at akka.dispatch.Mailbox.run(Mailbox.scala:221)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:231)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.pollAndExecAll(ForkJoinPool.java:1253)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1346)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

But I think that this is akka's fault, not ours.

rmetzger commented:
I'm done with my initial review.
If you have a minute @mxm, it would be good if you could check the CliFrontend changes, to see if they fit the architecture well.

rmetzger commented:
According to the design document, netty authentication is also part of this JIRA. Why was it not addressed?

vijikarthi commented:
>
According to the design document, netty authentication is also part of this JIRA. Why was it not addressed?

The netty layer is addressed as part of web layer authentication (T2-3 & T2-5 combined). Do you see a need to address this to some other part of the code as well?

rmetzger commented:
T2-3 is not about the web interface netty, its about the data transfer netty 
In Flink, we are using netty for (at least) three things:
- Akka is using Netty. This is addressed in the pull request
- The web interface is using Netty. Addressed as well 
- The user data (datastreams, etc.) is transferred using Netty between the TaskManagers as well.

vijikarthi commented:
>
T2-3 is not about the web interface netty, its about the data transfer netty
In Flink, we are using netty for (at least) three things:
    - Akka is using Netty. This is addressed in the pull request
    - The web interface is using Netty. Addressed as well
    - The user data (datastreams, etc.) is transferred using Netty between the TaskManagers as well.

Thanks Robert for the clarification. It was a good catch and I have clearly missed the Netty part used between TMs. I will create another JIRA and address the issue as part of it. 

I have incorporated your feedback comments and rebased the commits to latest master. Please retest and see if the patch is okay?

rmetzger commented:
Thank you for addressing my comments.

I've looked into the design document [1] again and some details are missing there. In particular its not clearly specified how and where the shared secrets are created and transferred in the cluster.
How is the secret transferred to the TaskManagers on YARN?
Is the using the  `JobManagerMessages.getRequestBlobManagerSecureCookie()` message always secure?
I think its important for security-related changes to first have a clear, documented security concept, and then an implementation that matches the documentation.
Maybe it also makes sense to start adding a page into the internals documentation section, explaining how the secure cookie is implemented.

@StephanEwen can you maybe add your thoughts here?

[1] https://docs.google.com/document/d/1-GQB6uVOyoaXGwtqwqLV8BHDxWiMO2WnVzBoJ8oPaAs/edit

vijikarthi commented:
>
How is the secret transferred to the TaskManagers on YARN?

Cookie is transferred to TM container through container environment variable and further gets populated to in-memory Flink configuration instance. The secure cookie is vulnerable (so as the Keytab file) to the users who has access to the container local resource storage area and that's the limitation we may have to deal with.

>Is using the JobManagerMessages.getRequestBlobManagerSecureCookie() message always secure?

I believe it is safe since Akka endpoints are secured using the shared token (cookie) and for someone to access the cookie using the "JobManagerMessages.getRequestBlobManagerSecureCookie()", they should have been authenticated first.

>
Maybe it also makes sense to start adding a page into the internals documentation section, explaining how the secure cookie is implemented.

I am planning to add a separate page to internals documentation explaining how the shared secret stuff is implemented for various deployment options - Standalone, Yarn, Mesos (tbd)

vijikarthi commented:
@rmetzger I have added internals documentation section and provided details on how secure cookie is implemented. I will address the missing Netty data transfer secure cookie part in FLINK-4635. Please take a look.

vijikarthi commented:
@rmetzger @StephanEwen are you guys waiting for any inputs from my side?

StephanEwen commented:
I think we are waiting for @rmetzger to follow up. He is out of office for a few days...

vijikarthi commented:
Thanks @StephanEwen

vijikarthi commented:
Addressed [FLINK-4635] Netty data transfer authentication (missing piece of FLINK-3930)

vijikarthi commented:
@rmetzger can you please take a look at the updated patch

rmetzger commented:
I'll take a look.

vijikarthi commented:
Resolved merge conflicts and squashed commits to rebase with master

mxm commented:
Thank you for your work so far @vijikarthi. I'll take a final look before we merge this.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestProtocol.java:
```diff
@@ -30,16 +30,29 @@
 		this.partitionProvider = partitionProvider;
 		this.taskEventDispatcher = taskEventDispatcher;
 		this.networkbufferPool = networkbufferPool;
+		this.secureCookie = (secureCookie == null || secureCookie.length() == 0) ? "": secureCookie;
```
See above comments about this.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestProtocol.java:
```diff
@@ -30,16 +30,29 @@
+		 * Client decoder does not validate the secure cookie from server since
+		 * the server protocol does not transmit the secure cookie on the wire
+		 */
+		clientMessageDecoder = new NettyMessage.NettyMessageDecoder(null);
```
But it still sends an empty cookie?

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -170,7 +176,8 @@ else if (currentPartitionQueue.isReleased()) {
 					else {
-						BufferResponse resp = new BufferResponse(buffer, currentPartitionQueue.getSequenceNumber(), currentPartitionQueue.getReceiverId());
+						BufferResponse resp = new BufferResponse(buffer, currentPartitionQueue.getSequenceNumber(),
+								currentPartitionQueue.getReceiverId());
```
just reformatting, no code changes (but i still have to check).

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -215,7 +222,8 @@ private void handleException(Channel channel, Throwable cause) throws IOExceptio
 		if (channel.isActive()) {
-			channel.writeAndFlush(new NettyMessage.ErrorResponse(cause)).addListener(ChannelFutureListener.CLOSE);
+			channel.writeAndFlush(new NettyMessage.ErrorResponse(cause))
+					.addListener(ChannelFutureListener.CLOSE);
```
just reformatting, no code changes (but i still have to check).

mxm commented on flink-runtime/src/main/scala/org/apache/flink/runtime/akka/AkkaUtils.scala:
```diff
@@ -389,7 +396,22 @@ object AkkaUtils {
+         |  }
+         |}
+         """.stripMargin
+    }else{
```
No need for the else block. We set set `require-cookie` to `off` in case it shouldn't be used

mxm commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -1008,6 +1008,9 @@ class JobManager(
       sender ! decorateMessage(libraryCacheManager.getBlobServerPort)
 
+    case RequestBlobManagerSecureCookie =>
+      sender ! decorateMessage(libraryCacheManager.getSecureCookie)
```
That is not a good idea for security!

mxm commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientSecureTest.java:
```diff
@@ -0,0 +1,78 @@
+/*
```
There is already a SSL test. BlobClientCookieTest seems more appropriate.

mxm commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobClientTest.java:
```diff
@@ -48,19 +49,18 @@
+	protected static BlobServer BLOB_SERVER;
 
-	/** The blob service client and server configuration */
-	private static Configuration blobServiceConfig;
```
These renamings are not necessary. Please revert.

mxm commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerDeleteTest.java:
```diff
@@ -18,6 +18,7 @@
 
 package org.apache.flink.runtime.blob;
 
+import org.apache.flink.configuration.ConfigConstants;
```
Unused import

mxm commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerGetTest.java:
```diff
@@ -18,6 +18,7 @@
 
 package org.apache.flink.runtime.blob;
 
+import org.apache.flink.configuration.ConfigConstants;
```
Unused import

mxm commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobServerPutTest.java:
```diff
@@ -18,6 +18,7 @@
 
 package org.apache.flink.runtime.blob;
 
+import org.apache.flink.configuration.ConfigConstants;
```
Unused import

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -108,6 +111,11 @@
+	private static final String fileName = "yarn-app.ini";
+	private static final String cookieKey = "secureCookie";
+
+	private final Option SECURE_COOKIE_OPTION;
```
This option can be retrieved from the main options. No need for yarn specific option. Please remove.

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -108,6 +111,11 @@
 	private final Options ALL_OPTIONS;
 
+	private static final String fileName = "yarn-app.ini";
+	private static final String cookieKey = "secureCookie";
```
Please see Yarn properties which should be used for this purpose: `writeYarnProperties(..)`.

Please no new configuration file.

mxm replied:
Good point. The yarn properties file should not be in the temp folder. Could you change that? We could do that in a separate commit.

If you want to store multiple application states, then let's change the layout of the yarn properties file. The goal should be to have only one configuration file and format. Could you do that in a separate commit? It shouldn't be too much work. Let me know if that would be possible.

vijikarthi replied:
Yes, I will make the change. 
- Do you object to retain the ini file format and port the current properties file implementation to INI format (to persist multiple application states)?

- Per current implementation (retrieveCluster), the CLI code fetches the application ID from properties file if not supplied through CLI argument. When we support multiple application state, then we expect application ID to be supplied always since there could be more than one application ID and the default functionality will go away. Do you concur? 

>
If we really need to provide backward compatibility support, then we could return the application ID from the INI file should there be just only instance persisted? If more than one application ID exists, then we throw an error indicating "Application ID" needs to be supplied as CLI argument.

Please let me know how you want me to approach and I will make the changes accordingly.

mxm replied:
I think the ini file format is actually fine. Could also be JSON but I don't mind. To not break backwards-compatibility, I think we have to keep the behavior to use the last-used application id in case none is supplied. We could have an extra config entry for that.

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -442,8 +453,10 @@ public static void runInteractiveCli(YarnClusterClient yarnCluster, boolean read
 						case "stop":
 							yarnCluster.shutdownCluster();
+							if (yarnCluster.hasBeenShutdown()) {
+								removeAppState(applicationId);
```
Cleanup is handled in the shutdown method. Please remove this. The cluster client is always in state running if the user calls stop.

vijikarthi replied:
Will move the logic to shutdown handler code

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -607,6 +680,18 @@ public int run(String[] args) {
+				LOG.debug("Secure cookie is provided as CLI argument and will be used");
+				yarnDescriptor.getFlinkConfiguration().setBoolean(ConfigConstants.SECURITY_ENABLED, true);
+				yarnDescriptor.getFlinkConfiguration().setString(ConfigConstants.SECURITY_COOKIE, secureCookieArg);
+			}
```
So much duplicate code here. Could be extracted to a method to avoid that.

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -703,6 +796,93 @@ public static File getYarnPropertiesLocation(Configuration conf) {
 		return new File(propertiesFileLocation, YARN_PROPERTIES_FILE + currentUser);
 	}
 
+	public static void persistAppState(String appId, String cookie) {
```
Please use the Yarn properties file for that.

vijikarthi commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -57,24 +61,37 @@
 		buffer.writeInt(HEADER_LENGTH + length);
 		buffer.writeInt(MAGIC_NUMBER);
+
+		buffer.writeInt(secureCookie.length());
```
Good catch. Will change it.

vijikarthi commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -112,9 +129,9 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+		// | FRAME LENGTH (4) | MAGIC NUMBER (4) | COOKIE (4) | ID (1) || CUSTOM MESSAGE |
+		// +------------------+------------------+------------+--------++----------------+
 		static LengthFieldBasedFrameDecoder createFrameLengthDecoder() {
 			return new LengthFieldBasedFrameDecoder(Integer.MAX_VALUE, 0, 4, -4, 4);
```
We don't strip the cookie and hence there is no change to adjust the decoder

vijikarthi commented:
Thanks @mxm for the review. I will incorporate your feedback and attach the patch.

>
When security is enabled, encryption should also be turned on by default. Otherwise we will transmit plain-text passwords over the wire.

Yes it makes sense. I will make a conditional check and throw an error if encryption is not enabled when security is enabled?

>
Should there be too modes for network transmission, 1) with cookie, one without? Do we need 32 bits for the cookie length? We should be precise about the maximum length. I saw it is set to 1024 in other places.

Yes, max cookie length validation is 1024. I will change the code where the buffer length was allocated to a high value, instead it will use the byte array length read from the message. 

> 
Should we really always send the cookie for every message? The security document mentions in T2-3 that we only want to authorize upon the first connection.

Yes, we took the approach to pass secure cookie for every message to keep minimal changes to the current design

>
Why do we transmit the cookie to the client? This seems like a major security concern. The client should always provide the cookie.
    edit: I see this has been specified in the document in T2-4. Still, I wonder if it would make sense to simply add this now because the workaround to fetch the cookie from the JobManager doesn't look appealing.

Good catch. I forgot to revert the code after the merge and it is not required. Will fix it. 

>
You added a Yarn specific cookie option which should be part of the general options instead.

It is added since secure cookie can be supplied when using both Yarn session CLI as well as Flink CLI. I have provided detailed explanation in one of the comments.

>
You've introduce a new config file to persist the app state. We already have the Yarn properties file for that.

I have provided explanation on why we need new config file in one of the comments. Please take a look.

vijikarthi commented:
@mxm - Please take a look when you get a chance?

mxm commented:
@vijikarthi I haven't forgotten about your PR. Thanks for the feedback. I'll get back to you today.

mxm commented:
CC @uce to check out the network layer changes. This is a very sensitive and performance critical part of Flink. We should be very sure nothing breaks it with the changes.

@vijikarthi Please have a look at the null checks in the network code. I would replace them with `checkNotNull` and never pass any null values in there. It would be desirable that turned off security doesn't have any overhead with the security support built in.

vijikarthi commented:
Addressed multiple application support/Yarn configuration file changes as part of FLINK-4950 patch.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -64,26 +64,25 @@
-	private static ByteBuf allocateBuffer(ByteBufAllocator allocator, byte id, String secureCookie, int length) {
+	private static ByteBuf allocateBuffer(ByteBufAllocator allocator, byte id, int length, String secureCookie) {
+		final Charset DEFAULT_CHARSET = Charset.forName("utf-8");
 		secureCookie = secureCookie == null ? "": secureCookie;
```
Simpler:
```java
if (secureCookie == null) {
    secureCookie = NO_SECURE_COOKIE;
}
```

mxm replied:
Still think we should never pass a null cookie. Then the check wouldn't be necessary.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -120,6 +136,15 @@ static LengthFieldBasedFrameDecoder createFrameLengthDecoder() {
+		final byte[] secureCookie;
+
+		public NettyMessageDecoder(String secureCookie) {
+			secureCookie = secureCookie == null ? "": secureCookie;
```
Should be removed.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServerConnection.java:
```diff
@@ -101,6 +102,14 @@ public void run() {
 
 			while (true) {
+
+				int keyLength = inputStream.read();
```
Here the cookie length is limited to one byte.

mxm commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -54,24 +58,36 @@
-	private static ByteBuf allocateBuffer(ByteBufAllocator allocator, byte id, int length) {
+	private static ByteBuf allocateBuffer(ByteBufAllocator allocator, byte id, int length, String secureCookie) {
+		final Charset DEFAULT_CHARSET = Charset.forName("utf-8");
+		secureCookie = secureCookie == null ? "": secureCookie;
```
Should be removed in favor of never passing a null value here.

mxm commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -788,75 +719,125 @@ private void logAndSysout(String message) {
+		return null;
+	}
+
+	public static YarnAppState retrieveYarnAppState(String appId) {
```
You could pass the config here which would avoid re-creating it in this method.

vijikarthi commented:
@mxm - Sorry that I have missed to address some of your comments. Attached patch that includes Netty code null precondition validation and fixes the Blob service cookie length issue. Please take a look and see if they are okay? Thanks for your patience.

mxm commented:
Thank you for the changes. I wonder, could we remove the cookie header completely for Netty or the BlobServer in case the authorization is turned off? The Netty protocol has a `MAGIC_NUMBER` which is checked when decoding the message. We could use a different "magic number" to check whether we use the normal or the cookie-based Netty protocol. This would eliminate all the overhead of the cookie transmission. Furthermore, we should strip the cookie from the message once we have verified it is correct.

StephanEwen commented on flink-core/src/main/java/org/apache/flink/configuration/GlobalConfiguration.java:
```diff
@@ -144,8 +144,15 @@ private static Configuration loadYAMLResource(File file) {
+
+					//to prevent logging the secure cookie
+					if(key.equals(ConfigConstants.SECURITY_COOKIE) && value != null) {
+						value = "******";
```
Can all suppressed keys have a common prefix?

This should also be guarded by a test ;-)

StephanEwen commented on flink-dist/src/main/resources/flink-conf.yaml:
```diff
@@ -173,3 +174,13 @@ jobmanager.web.port: 8081
 # zookeeper.sasl.service-name: zookeeper
+
+#==============================================================================
+# Service Authorization Configuration (optional configuration)
```
Can we group all security related parts together?

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java:
```diff
@@ -57,24 +61,37 @@
 	// ------------------------------------------------------------------------
 
-	static final int HEADER_LENGTH = 4 + 4 + 1; // frame length (4), magic number (4), msg ID (1)
+	static final int HEADER_LENGTH = 4 + 4 + 4 + 1; // frame length (4), magic number (4), Cookie (4), msg ID (1)
```
Looks like this is the cookie length

StephanEwen commented:
The Netty logic needs some improvements:

  - The cookie is added to every single message/buffer that is transferred. That is too much - securing the integrity of the stream is responsibility of the encryption layer. The cookie should be added to requests messages that establish connections only.

  - Charset lookups and cookie to bytes encoding happens for every buffer, rather than once in an initialization step.

  - The String to byte conversion is not consistent. Sometimes it uses the default platform encoding, sometimes "UTF-8".

vijikarthi commented:
>
The cookie is added to every single message/buffer that is transferred. That is too much - securing the integrity of the stream is responsibility of the encryption layer. The cookie should be added to requests messages that establish connections only.

I will change the code to address cookie handling right after the SSL handshake using a new handler and drop the cookie passing for every messages. The handler will be added to the pipeline of both `NettyClient` and `NettyServer`. Client will send the cookie when the channel becomes active and the server will validate and keep track of the clients that are authorized. 

Here is the pseudo-code for Client and Server handlers. Please take a look and let me know if you are okay with this approach and I will modify the code.

---
	public static class ClientCookieHandler extends ChannelInboundHandlerAdapter {

		private final String secureCookie;

		final Charset DEFAULT_CHARSET = Charset.forName("utf-8");

		public ClientCookieHandler(String secureCookie) {
			this.secureCookie = secureCookie;
		}


		@Override
		public void channelActive(ChannelHandlerContext ctx) throws Exception {
			super.channelActive(ctx);

			if(this.secureCookie != null && this.secureCookie.length() != 0) {
				final ByteBuf buffer = Unpooled.buffer(4 + this.secureCookie.getBytes(DEFAULT_CHARSET).length);
				buffer.writeInt(secureCookie.getBytes(DEFAULT_CHARSET).length);
				buffer.writeBytes(secureCookie.getBytes(DEFAULT_CHARSET));
				ctx.writeAndFlush(buffer);
			}
		}
	}

	public static class ServerCookieDecoder extends MessageToMessageDecoder<ByteBuf> {

		private final String secureCookie;

		private final List<Channel> channelList = new ArrayList<>();

		private final Charset DEFAULT_CHARSET = Charset.forName("utf-8");

	       public ServerCookieDecoder(String secureCookie) {
			this.secureCookie = secureCookie;
		}

		@Override
		protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) throws Exception {

			if(secureCookie == null || secureCookie.length() == 0) {
				return;
			}

			if(channelList.contains(ctx.channel())) {
				return;
			}

			//read cookie based on the cookie length passed
			int cookieLength = msg.readInt();
			if(cookieLength != secureCookie.getBytes(DEFAULT_CHARSET).length) {
				String message = "Cookie length does not match with source cookie. Invalid secure cookie passed.";
				throw new IllegalStateException(message);
			}

			//read only if cookie length is greater than zero
			if(cookieLength > 0) {

				final byte[] buffer = new byte[secureCookie.getBytes(DEFAULT_CHARSET).length];
				msg.readBytes(buffer, 0, cookieLength);

				if(!Arrays.equals(secureCookie.getBytes(DEFAULT_CHARSET), buffer)) {
					LOG.error("Secure cookie from the client is not matching with the server's identity");
					throw new IllegalStateException("Invalid secure cookie passed.");
				}

				LOG.info("Secure cookie validation passed");

				channelList.add(ctx.channel());
			}

		}
	}
---

vijikarthi commented:
@StephanEwen @mxm - Could you please review the proposed change and let me know if you are okay with it.

vijikarthi commented on docs/internals/flink_security.md:
```diff
@@ -84,4 +86,79 @@ Security implementation details are based on <a href="https://github.com/apache/
+
+The secure cookie will be made available as container environment variable for the application containers (JM/TM) to make use of it.
+
+On the client machine from where the Yarn session CLI is used to create the Flink application, the application specific secure cookie will be persisted in an INI file format in the user home directory. Any subsequent access to the Flink cluster using Yarn Session CLI (by passing the application ID) will automatically include appropriate secure cookie associated with the application ID to communic ...
```
No. This happens for both user provided as well as auto-generated scenario

vijikarthi commented:
@StephanEwen, @mxm  I have updated the documentation changes as suggested, moved common code from BlobUtils to SecurityContext, added new ConfigOptions class for security configurations lookup. 

>
The cookie is added to every single message/buffer that is transferred. That is too much - securing the integrity of the stream is responsibility of the encryption layer. The cookie should be added to requests messages that establish connections only.

I have added a new handler code to front load the secure cookie validation. These handlers are added to both `NettyServer` and `NettyClient` pipeline right after the SSL handler is added. I still kept the original code that passes cookie for every message (will remove the logic if you are okay with the handler implementation)

Please review and let me know your feedback.

Rucongzhang commented:
@vijikarthi ,when you will push this issue to the master? I can help you, if you need any help. Thanks!

StephanEwen commented:
Sorry for chiming in a bit late here with this more fundamental question.

I would like to understand from a security architecture, what additional security this shared secret gives us:
  - If there is no encryption, then this shared secret is not very secure, as it can be sniffed from the network
  - When there is encryption, isn't the current assumption that all parties have access to the server-side certificate? Would that already be a form of shared secret, meaning that certificate-based authentication as part of the SSL handshake already covers the mechanism of a shared secret?

vijikarthi commented:
@StephanEwen The shared secret serves can be considered as an additional security extension on top of TLS integration, thus it designates only an authorized identity to execute actions on a running cluster.

EronWright commented:
@StephanEwen keep in mind that Flink's current SSL support in Flink doesn't achieve _mutual authentication_ - there's no client certificate there.    With SSL enabled, an untrusted client can launch jobs in your Flink cluster and thus gain access to the Kerberos credential associated with the cluster.

SSL mutual authentication is a good alternative to a shared secret, but at the time we were limited to built-in Akka functionality (which doesn't include mutual auth).   Given the "flakka" fork that's now in place, a pure SSL solution might now be possible (I haven't thought it through completely).

The fact remains that, today, _all the secrets known to a Flink job are exposed to everyone who can connect to the cluster's endpoint_.  

It would be nice to construct a holistic plan that worked out how the Web UI would support authentication and also incorporated FLIP-6.      Both YARN and Mesos interpose a web proxy for the UI with its own limitations, notably no support for SSL mutual auth.

StephanEwen commented:
Thanks Eron, that makes a lot of sense.

My first thought would be: Let's add SSL mutual authentication. That seems desirable anyways and we would not need another mechanism (shared secret). Do you know if newer versions of Akka support this mutual auth? We may be able to upgrade if we drop Java 7, or we could see if there is a lightweight way to patch this into flakka.

StephanEwen commented:
@vijikarthi I hope you are okay with exploring that option - this is not saying that this pull request is not a good solution, but whenever we have to maintain less code it makes things easier.

WangTaoTheTonic commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CookieHandler.java:
```diff
@@ -0,0 +1,130 @@
+
+		private final String secureCookie;
+
+		private final List<Channel> channelList = new ArrayList<>();
```
Is it better to use `Set` instead of a `List` here? As it is mainly used for lookup.

WangTaoTheTonic commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CookieHandler.java:
```diff
@@ -0,0 +1,130 @@
+
+				LOG.info("Secure cookie validation passed");
+
+				channelList.add(ctx.channel());
```
It seems like we only have channels added, but not removed. Would this list increase indefinitely?

vijikarthi commented:
@StephanEwen It's absolutely fine with me and I will cancel this PR.

EronWright commented:
Note to future self: to generate a self-signed certificate, use `CertAndKeyGen` and see [OPENDJ-2247](https://bugster.forgerock.org/jira/browse/OPENDJ-2247).
